File: 백준/레벨Gold III/1005. ACM Craft/ACM Craft.cc
#include <bits/stdc++.h>
using namespace std;

int N, K, W;
vector<int> buildTime;
vector<vector<int>> graph;
vector<int> inDegree;

vector<int> topologicalSort() {
    deque<int> stk;
    vector<int> dp(N + 1, 0);

    for (int i = 1; i <= N; i++) {
        if (inDegree[i] == 0) {
            stk.push_back(i);
            dp[i] = buildTime[i];
        }
    }

    while (!stk.empty()) {
        int u = stk.front();
        stk.pop_front();

        for (int v : graph[u]) {
            inDegree[v]--;
            dp[v] = max(dp[v], dp[u] + buildTime[v]);

            if (inDegree[v] == 0) {
                stk.push_back(v);
            }
        }
    }

    return dp;
}

int test() {
  // 빌드타임 초기화  
  cin >> N >> K;

  buildTime.assign(N + 1, 0);

  for (int i = 1; i <= N; i++) {
    cin >> buildTime[i];
  }

  // 인접리스트, 진입차수 초기화
  graph.assign(N + 1, vector<int>());
  inDegree.assign(N + 1, 0);

  while (K--) {
    int u, v;
    cin >> u >> v;

    graph[u].push_back(v);
    inDegree[v]++;
  }

  // 승리 조건
  cin >> W;

  // 위상 정렬
  vector<int> dp = topologicalSort();

  return dp[W];
}

int main(void) {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int T;
  cin >> T;

  while (T--) {
    cout << test() << "\n";
  }
}

File: 백준/레벨Gold III/1005. ACM Craft/ACM Craft.java
import java.io.*;
import java.util.*;

class Solution {
    static int nextInt() throws IOException {
        int c, n = System.in.read() & 15; // 1의 자리 n, 다음 자리 c
        while((c = System.in.read()) > 32) { // 탭 9, 개행 10, 공백 32
            n = n * 10 + (c & 15); // 한자리씩 올리기
        }
        return n;
    }

    static int N, K, W;
    static int[] buildTime;
    static int[] inDegree;
    static List<List<Integer>> graph;

    private int[] topologicalSort() {
        Deque<Integer> stk = new ArrayDeque<>();
        int[] dp = new int[N + 1];

        for (int i = 1; i <= N; i++) {
            if (inDegree[i] == 0) {
                stk.push(i);
                dp[i] = buildTime[i];
            }
        }

        while (!stk.isEmpty()) {
            int u = stk.pop();

            for (int v : graph.get(u)) {
                inDegree[v]--;
                dp[v] = Math.max(dp[v], dp[u] + buildTime[v]);

                if (inDegree[v] == 0) {
                    stk.push(v);
                }
            }
        }

        return dp;
    }

    private int test() throws IOException {
        // 빌드타임 초기화
        N = nextInt();
        K = nextInt();
        buildTime = new int[N + 1];

        for (int i = 1; i <= N; i++) {
            buildTime[i] = nextInt();
        }

        // 인접리스트, 진입차수 초기화
        graph = new ArrayList<>();
        inDegree = new int[N + 1];
        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }

        while (K-- > 0) {
            int u = nextInt();
            int v = nextInt();
            graph.get(u).add(v);
            inDegree[v]++;
        }

        // 승리 조건
        W = nextInt();

        // 위상 정렬
        int[] dp = topologicalSort();

        return dp[W];
    }

    public void solution() throws IOException {
        int T = nextInt();
        StringBuilder sb = new StringBuilder();

        for (int t = 0; t < T; t++) {
            sb.append(test()).append("\n");
        }

        System.out.println(sb);
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        new Solution().solution();
    }
}